### ISSUE TITLE
Choose GPU device

### ISSUE THREAD

**Issue by a2018c** [ikhlestov/vision_networks#24]
Hi,

Thanks a lot for this "vision_networks"!

I have two GPUs in my machine, exactly same type of GPUs. The first one is also used to connect to monitors, I found that the first GPUs has less available RAM than the second one. 

Found device 0 with properties: 
name: GeForce GTX 1070 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.683
pciBusID: 0000:01:00.0
totalMemory: 7.92GiB freeMemory: 5.14GiB

Found device 1 with properties: 
name: GeForce GTX 1070 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.683
pciBusID: 0000:02:00.0
totalMemory: 7.92GiB freeMemory: 7.70GiB

Two questions:
First, is there a way to use two GPUs in parallel to speed up the training process? Here, it always uses 
the first GPU and the second one is idle.

Second, is there a way to choose which GPU to be used if it only supports one GPU?

Regards

**Comment by ikhlestov** [comment]
Hi!

Regarding parallel training - not, it was not implemented, because I have only one GPU).

If you want to choose some exact GPU device you may provide environment flag while running the script like `export CUDA_VISIBLE_DEVICES='0' && python train.py`. With this only first GPU will be available for script. You may hide all gpu devices with `CUDA_VISIBLE_DEVICES=''` (empty quotes) flag.

**Comment by a2018c** [comment]
I tried "--growth_rate=12 --depth=250", I got: Time per epoch: 3:07:27.282798, Est. complete in: 38 days, 22:08:57

That is why I want to use two 2 GPUs. 

I tried to modify dense_net.py with following changes, one of the GPU loading is still low

###
def train_one_epoch(self, data, batch_size, learning_rate, num_gpus):
        num_examples = data.num_examples
        total_loss = []
        total_accuracy = []

        for j in range(num_examples // batch_size):
          ### ADD (I have 2 GPUs, and trying to make sure GPU:1 first
          for d in ['/device:GPU:1', '/device:GPU:0']:
            with tf.device(d):
            ###
               batch = data.next_batch(batch_size)
               images, labels = batch
               feed_dict = {
                   self.images: images,
                   self.labels: labels,
                   self.learning_rate: learning_rate,
                   self.is_training: True,
               }
               fetches = [self.train_step, self.cross_entropy, self.accuracy]

               result = self.sess.run(fetches, feed_dict=feed_dict)

               _, loss, accuracy = result
               total_loss.append(loss)
               total_accuracy.append(accuracy)

               if self.should_save_logs:
                   self.batches_step += 1
                   self.log_loss_accuracy(
                       loss, accuracy, self.batches_step, prefix='per_batch',
                       should_print=False)

        mean_loss = np.mean(total_loss)
        mean_accuracy = np.mean(total_accuracy)
        return mean_loss, mean_accuracy
###

Is my approach not correct? Please help

**Comment by ikhlestov** [comment]
As I remember I didn't use any hardcoded GPU allocation.. I can advice you to try same device placements but with some tiny network and maybe even dummy images(np.random or like this)

**Comment by a2018c** [comment]
export CUDA_VISIBLE_DEVICES='1' works! Thanks

It takes  38 days to train  "--growth_rate=12 --depth=250" which is too long.  Could you help to advise if my  modification (above) is correct?

I have to set batch_size to 4 for "--growth_rate=12 --depth=250  --dataset=C100"

+-----------------------------------------------------------------------------+
| NVIDIA-SMI 387.34                 Driver Version: 387.34                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 107...  Off  | 00000000:01:00.0  On |                  N/A |
| 28%   59C    P2    38W / 180W |   1646MiB /  8112MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 107...  Off  | 00000000:02:00.0 Off |                  N/A |
| 45%   65C    P2   121W / 180W |   7877MiB /  8114MiB |     95%      Default |
+-------------------------------+----------------------+----------------------+

The loading of "gpu:0" is close to idle

**Comment by ikhlestov** [comment]
At least I will advice you to use default depth == 40. Depth == 250 should be used for the imageNet dataset, which is not provided out of the box.

**Comment by a2018c** [comment]
'--growth_rate=12 --depth=250  --dataset=C100', Time per epoch: 3:25:13.253673, Est. complete in: 42 days, 11:15:49

Need advices how to speed up this

**Comment by ikhlestov** [comment]
If you want use depth == 250, which was designed for ImageNet, and you want to speedup the training - you should implement multi GPU training yourself and train such network on multi GPU instances(local or remote). For example with 8 GPU AWS instances. Because it's really large network and of course it will take a lot of time for such training.
ps. Also I think you may rewrite data loader to some tensorflow based implementation.

**Comment by a2018c** [comment]
I got error:

2018-02-21 10:55:02.077908: W tensorflow/core/common_runtime/bfc_allocator.cc:277] ****************************************************************************************************
2018-02-21 10:55:02.077926: W tensorflow/core/framework/op_kernel.cc:1198] Resource exhausted: OOM when allocating tensor with shape[3,3,9520,40] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc

Is there a way in configuration set to the RAM limit and avoid OOM?

**Comment by ikhlestov** [comment]
because tensorflow use static graph - it build whole network at once. So it's impossible to reduce memory consumption without reducing network size. You may only reduce network size itself or update your GPU.
